hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,740 @@
+	SetThreadNameInt(m_thread->native_handle());
+
+	sOnConnected(this);
+	m_bIsStarted = true;
+	return (m_thread != NULL);
+}
+bool VirtualThermostat::StopHardware()
+{
+	m_stoprequested = true;
+	{
+		RequestStop();
+		m_thread->join();
+		m_thread.reset();
+	}
+	m_bIsStarted = false;
+	return true;
+}
+void VirtualThermostat::Do_Work()
+{
+	int sec_counter = 0;
+	_log.Log(LOG_STATUS, "VirtualThermostat: Worker started...");
+	while (IsStopRequested(1000) == false)
+	{
+		if (m_stoprequested)
+			break;
+		sec_counter++;
+
+		if (sec_counter % 12 == 0) {
+			m_LastHeartbeat = mytime(NULL);
+		}
+		time_t atime = mytime(NULL);
+		struct tm ltime;
+		localtime_r(&atime, &ltime);
+#if 1
+        if (ltime.tm_min != m_ScheduleLastMinute)
+		{
+			ScheduleThermostat(ltime.tm_min);
+			m_ScheduleLastMinute = ltime.tm_min;
+		}
+#else
+        //test : call every 6 secconds
+        if ( ltime.tm_sec %6 ==0 )
+		{
+			ScheduleThermostat(ltime.tm_sec/6);
+			m_ScheduleLastMinute = ltime.tm_sec;
+		}
+#endif
+	}
+	_log.Log(LOG_STATUS, "VirtualThermostat: Worker stopped...");
+}
+bool VirtualThermostat::WriteToHardware(const char *pdata, const unsigned char length)
+{
+	return true;
+}
+
+//time for the power time modulation in minute
+#define MODULATION_DURATION 10
+//  number of step in percent %
+#define MODULATION_STEP     10
+
+//guve the output power swicth value in the time from the modulation percent
+const short ThermostatOutput[MODULATION_STEP+1][MODULATION_DURATION]={
+//            0 1 2 3 4 5 6 7 8 9    
+/* 000 % */ { 0,0,0,0,0,0,0,0,0,0 },
+/* 010 % */ { 1,0,0,0,0,0,0,0,0,0 },
+/* 020 % */ { 1,1,0,0,0,0,0,0,0,0 },
+/* 030 % */ { 1,1,1,0,0,0,0,0,0,0 },
+/* 040 % */ { 1,1,1,1,0,0,0,0,0,0 },
+/* 050 % */ { 1,1,1,1,1,0,0,0,0,0 },
+/* 060 % */ { 1,1,1,1,1,1,0,0,0,0 },
+/* 070 % */ { 1,1,1,1,1,1,1,0,0,0 },
+/* 080 % */ { 1,1,1,1,1,1,1,1,0,0 },
+/* 090 % */ { 1,1,1,1,1,1,1,1,1,0 },
+/* 100 % */ { 1,1,1,1,1,1,1,1,1,1 },
+                    
+};                  
+
+//compute the thermostat output switch value 
+// the output is modulated in a time period of 10 minute
+// each minute , the output is activated dependeing time and percent.
+//input : Min : the minute counter 0..59
+//input : PowerPercent : the power modulation  0..100%
+//output 1 : activated
+
+short VirtualThermostat::ComputeThermostatOutput ( int Min , int PowerPercent )
+{
+  if (PowerPercent>100) PowerPercent=100;
+  //get row value for ThermostatOutput
+  PowerPercent = (PowerPercent  ) / MODULATION_STEP ;
+  short switchValue = ThermostatOutput[PowerPercent][Min%MODULATION_DURATION]; 
+  return switchValue;
+}                    
+  
+//return the poqer modulation in function of Room , Exterior and Target Temperature,
+int VirtualThermostat::ComputeThermostatPower ( int index , double RoomTemp , double TargetTemp , double ExteriorTemp , double CoefProportional , double CoefIntegral , double CoefDerivated  )
+{
+    (void)CoefDerivated;
```

Comment:
I don't think this line should be here?

Reply:
it is to avoid unused argument warning

Reply:
Thats an option... the other option is to remove these arguments.. 😁